Accept Ethereum passphrase via file on disk rather than command line

mk-livepeer opened the pull request:
When bringing up nodes on an actual network, `-ethPassword` can be
specified on the command line to supply the Ethereum account's
passphrase and allow unattended startup and operations. This works
for now, but it is very insecure. A attacker can compromise a system,
run `ps aux` and then see all command lines, including the go-livepeer
command line containing the Ethereum passphrase in clear text.

This adds the ability for the user to supply a filepath rather than a
cleartext password to the `-ethPassword` parameter. If the supplied
string is a valid file path, the passphrase will be read from the file,
otherwise the supplied string will be used as-is.

closes #1256

- [x] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

mkrufky committed:
Accept Ethereum passphrase via file on disk rather than command line

When bringing up nodes on an actual network, `-ethPassword` can be
specified on the command line to supply the Ethereum account's
passphrase and allow unattended startup and operations. This works
for now, but it is very insecure. A attacker can compromise a system,
run `ps aux` and then see all command lines, including the go-livepeer
command line containing the Ethereum passphrase in clear text.

This adds the ability for the user to supply a filepath rather than a
cleartext password to the `-ethPassword` parameter. If the supplied
string is a valid file path, the passphrase will be read from the file,
otherwise the supplied string will be used as-is.

closes #1256

mkrufky committed:
Document func GetPass(s string, idx int) (string, error)

GetPass attempts to read a file for a password at the supplied location.
If it fails, then the original supplied string will be returned to the caller.

mkrufky committed:
getpass: return an error when supplied a directory

kyriediculous commented on common/getpass.go:
```diff
@@ -0,0 +1,46 @@
+// GetPass attempts to read a file for a password at the supplied location.
+// If it fails, then the original supplied string will be returned to the caller.
+// A valid string will always be returned, regardless of whether an error occurred.
+func GetPass(s string, idx int) (string, error) {
```
For now since we don't support unlocking multiple accounts I think we could just omit the `idx` argument and return the `0` index by default

mkrufky replied:
Since it's a utility function, I thought it would be a good idea to expose this now rather than later.  I could imagine someone running an OT node but want to use separate account for the trans....   Oh, that doesn't make sense :-P

If you don't think we'll ever need the feature, I can remove it.

mk-livepeer replied:
removed in eeb1a3eaf56ff60f903b8884df5f77a0c3602dcc

kyriediculous replied:
I don't see much reason for it in the near future in fact our `AccountManager` implementation only supports a single account as is. I'm not even sure why geth or parity expose the functionality to unlock multiple accounts, do you use multiple unlocks on a single node for mining ?

mkrufky replied:
I use multiple accounts on all of my mining nodes, yes.  There are *many* reasons why a systems integrator may associate multiple accounts to a geth instance.  I'm not arguing it here for our case though - I removed it in eeb1a3eaf56ff60f903b8884df5f77a0c3602dcc

mk-livepeer committed:
getpass_test: fix typo: Extsts --> Exists

kyriediculous commented on common/getpass.go:
```diff
@@ -0,0 +1,46 @@
+// GetPass attempts to read a file for a password at the supplied location.
+// If it fails, then the original supplied string will be returned to the caller.
+// A valid string will always be returned, regardless of whether an error occurred.
+func GetPass(s string, idx int) (string, error) {
```
What do you think of just moving this into `common/util.go` or `eth/helpers` instead of its own file?

mkrufky replied:
If you think that's best, we can do that...  I put it into its own file because I knew I would be writing tests and I prefer to keep things in separate files when each function may have multiple permutations to test.

kyriediculous replied:
Actuallly I thought about this some more and what we should do is the following

In livepeer.go , just pass `*ethPassword` to `client.Setup` https://github.com/livepeer/go-livepeer/blob/dc878b5f938b0c7c7fdf5e69bccf3122c78e78b4/cmd/livepeer/livepeer.go#L352

So no changes have to actually be made here (except perhaps altering the `-ethPassword` flag description). 

In `eth/accountmanager.go` there is an `Unlock()` function, this is where we also prompt for the passphrase in case the `-ethPassword` flag was not provided on node startup.  This is the method that's being called in `client.Setup()` to unlock the account. 

https://github.com/livepeer/go-livepeer/blob/dc878b5f938b0c7c7fdf5e69bccf3122c78e78b4/eth/client.go#L175-L182

My suggestion is that we therefore move this new helper you created  into `eth/helpers` so that we can use it in `eth/accountManager.Unlock()`. 

In addition we can change the current default behaviour of the password prompt to also be able to take in a path (this is where we will change from how geth does is). This means that the user can always supply a password or path regardless of whether the `-ethPassword` flag is used. 

https://github.com/livepeer/go-livepeer/blob/dc878b5f938b0c7c7fdf5e69bccf3122c78e78b4/eth/accountmanager.go#L80-L103

`AccountManager` really is our specific layer on top of what go-ethereum provides so I think it makes sense to implement it there instead of `cmd` or `common`. WDYT?

mkrufky replied:
I'm not sure I agree with this -- this same function will be used for `-orchSecret`.  With that in mind, does it still make sense to make the changes you're proposing?

kyriediculous replied:
> I'm not sure I agree with this -- this same function will be used for orchSecret.

Good point, did not think ahead here. We can leave it in `common` then but I'd have a slight preference having it in `util.go` then, but I'll leave that decision up to you.

As for the rest of the changes described I think it's worth to implement them, we can import the helper from the `common` package then. 

We can do something like 

```
func(am *accountManager) Unlock(passphrase string) error {
	if passphrase == "" {
		glog.Infof("Please enter the passphrase or path to password file to unlock Ethereum account %v", am.account.Address.Hex())
		var err error
		passphrase, err = getPassphrase(false)
		if err != nil {
			return err
		}
	}

	pass, err := common.GetPass(passphrase)
	if err != nil {
		return err
	}

	if err := am.keyStore.Unlock(am.account, pass); err != nil  {
		return err
	}

	am.unlocked = true

	glog.Infof("Unlocked ETH account: %v", am.account.Address.Hex())

	return nil
}
```

mkrufky replied:
Done.  Please check :-)

As far as moving `GetPass` into `common/util` , it's an anti-pattern.  It doesn't really go along with the spirit of [Idiomatic Golang](https://about.sourcegraph.com/go/idiomatic-go) ... On the other hand, neither do some of the other parts of this repository.

I hope we'll have a chance to clean up and refactor this sort of thing before too much time goes by.  As an open source project, we should strive to be more idiomatic.

Meanwhile, it's also important to stick with the style of existing code.  ...

It's clear how I feel about this, but I'll move it if that's your preference.

mkrufky replied:
I don't think it's necessary for the interactive prompt for passphrase to be able to read from a file.  The point of this was to be able to keep the cleartext passphrase out of the command line for security purposes.

kyriediculous replied:
> I don't think it's necessary for the interactive prompt for passphrase to be able to read from a file.

You're probably right.

Changes look good, can we clean up the commit history ? 

> As far as moving GetPass into common/util , it's an anti-pattern. It doesn't really go along with the spirit of Idiomatic Golang ... On the other hand, neither do some of the other parts of this repository.

I agree that some parts of the codebases aren't really diehard idomatic go, but as you say

> Meanwhile, it's also important to stick with the style of existing code.

mkrufky replied:
Best practice is to squash on merge.  There is no need to flatten the commits in the PR itself.

mkrufky committed:
getpass: dont support multiple passwords per file

mkrufky committed:
getpass: remove redundant check

mkrufky committed:
getpass_test: add separate tests for single line and multiline files

mkrufky committed:
Read passhprase from file within `accountManager.Unlock()`

Pass raw `*ethPassword` to `client.Setup()` and attempt to
read passhprase from file within `accountManager.Unlock()`
